src/scale_impls/tracing_decoder/error.rs

```diff
@@ -0,0 +1,198 @@
+        TraceDecodingError { inner: value }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum TraceDecodingErrorInner<Val> {
+    FromDecodeError(DecodeError),
+    FromVisitor(VisitorError<Val>),
+}
+
+#[derive(Clone, Debug)]
+struct VisitorError<Val> {
+    at: Path,
+    decoded_so_far: Val,
+    decode_error: DecodeError,
+}
+
+impl<Ctx: core::fmt::Debug> core::fmt::Display for TraceDecodingError<Value<Ctx>> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match &self.inner {
+            TraceDecodingErrorInner::FromDecodeError(e) => {
+                write!(f, "Error decoding value: {e}")
+            }
+            TraceDecodingErrorInner::FromVisitor(e) => {
+                write!(
+                    f,
+                    "Error decoding value at {}: {}\nDecoded so far:\n\n",
+                    e.at, e.decode_error,
+                )?;
+                display_value_with_typeid(f, &e.decoded_so_far)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<Ctx: core::fmt::Debug> std::error::Error for TraceDecodingError<Value<Ctx>> {}
+
+impl<TypeId> From<DecodeError> for TraceDecodingError<TypeId> {
+    fn from(value: DecodeError) -> Self {
+        TraceDecodingErrorInner::FromDecodeError(value).into()
+    }
+}
+
+impl<TypeId> From<codec::Error> for TraceDecodingError<TypeId> {
+    fn from(value: codec::Error) -> Self {
+        TraceDecodingErrorInner::FromDecodeError(value.into()).into()
+    }
+}
+
+fn display_value_with_typeid<Id: core::fmt::Debug>(
+    f: &mut core::fmt::Formatter<'_>,
+    value: &Value<Id>,
+) -> core::fmt::Result {
+    use crate::string_impls::{fmt_value, FormatOpts, Formatter};
+
+    let format_opts = FormatOpts::new()
+        .spaced()
+        .context(|type_id, writer: &mut &mut core::fmt::Formatter| write!(writer, "{type_id:?}"))
+        .custom_formatter(|value, writer| custom_hex_formatter(value, writer));
+    let mut formatter = Formatter::new(f, format_opts);
+
+    fmt_value(value, &mut formatter)
+}
+
+fn custom_hex_formatter<T, W: core::fmt::Write>(
+    value: &Value<T>,
+    writer: W,
+) -> Option<core::fmt::Result> {
+    // Print unnamed sequences of u8s as hex strings; ignore anything else.
+    if let ValueDef::Composite(Composite::Unnamed(vals)) = &value.value {
+        for val in vals {
+            if !matches!(val.value, ValueDef::Primitive(Primitive::U128(n)) if n < 256) {
+                return None;
+            }
+        }
+        Some(value_to_hex(vals, writer))
+    } else {
+        None
+    }
+}
+
+// Just to avoid needing to import the `hex` dependency just for this.
+fn value_to_hex<T, W: core::fmt::Write>(vals: &Vec<Value<T>>, mut writer: W) -> core::fmt::Result {
+    writer.write_str("0x")?;
+    for val in vals {
+        if let ValueDef::Primitive(Primitive::U128(n)) = &val.value {
+            let n = *n as u8;
+            writer.write_char(u4_to_hex(n >> 4))?;
```

Comment:
nit: can't this be also written like this? 
```rust
            writer.write_str(format!("{n:02X}").as_ref())?;
```

Reply:
It could, but `format!` is then allocating a little String for each byte, which I wanted to avoid :) Probably not a huge deal either way given this crate not being super performance focused or anything